src/mcp/index.ts

```diff
@@ -50,15 +53,53 @@ export class FirebaseMcpServer {
           mcp_client_version: clientInfo.version,
         });
       }
+      if (!this.clientInfo?.name) this.clientInfo = { name: "<unknown-client>" };
+
+      this._ready = true;
+      while (this._readyPromises.length) {
+        this._readyPromises.pop()?.resolve();
+      }
     };
-    this.projectRoot =
-      options.projectRoot ??
-      (configstore.get(PROJECT_ROOT_KEY) as string) ??
-      process.env.PROJECT_ROOT ??
-      process.cwd();
+    this.detectProjectRoot();
     this.detectActiveFeatures();
   }
 
+  /** Wait until initialization has finished. */
+  ready() {
+    if (this._ready) return Promise.resolve();
+    return new Promise((resolve, reject) => {
+      this._readyPromises.push({ resolve: resolve as () => void, reject });
+    });
+  }
+
+  private get clientConfigKey() {
+    return `mcp.clientConfigs.${this.clientInfo?.name || "<unknown-client>"}`;
+  }
+
+  getStoredClientConfig(): ClientConfig {
+    return configstore.get(this.clientConfigKey) || {};
+  }
+
+  updateStoredClientConfig(update: Partial<ClientConfig>) {
+    const config = configstore.get(this.clientConfigKey) || {};
+    const newConfig = { ...config, ...update };
+    configstore.set(this.clientConfigKey, newConfig);
+    return newConfig;
+  }
+
+  async detectProjectRoot(): Promise<string> {
+    await this.ready();
+    if (this.cachedProjectRoot) return this.cachedProjectRoot;
+    const storedRoot = this.getStoredClientConfig().projectRoot;
+    if (storedRoot) {
+      this.cachedProjectRoot = storedRoot;
+      return storedRoot;
+    }
```

Comment:
nit: the above if branch looks redundant. Already taken care of below

Reply:
You're right!